go/ql/test/library-tests/semmle/go/frameworks/Fasthttp/fasthttp.go

```diff
@@ -0,0 +1,206 @@
+package main
+
+// go:generate depstubber -vendor  "github.com/valyala/fasthttp" Args,Client,Cookie,FS,HostClient,LBClient,PathRewriteFunc,Request,RequestCtx,RequestHandler,RequestHeader,Response,ResponseHeader,Server,TCPDialer,URI,LBClient,PipelineClient AcquireURI,Serve,DialDualStack,Dial,DialTimeout,DialDualStackTimeout,Get,GetDeadline,GetTimeout,Post,Do,DoRedirects,AppendHTMLEscapeBytes,AppendHTMLEscape,AppendQuotedArg,ServeFileBytesUncompressed,ServeFileBytes,ServeFileUncompressed,ServeFile,SaveMultipartFile,DoTimeout,DoDeadline
+
+import (
+	"bufio"
+	"net"
+	"time"
+
+	"github.com/valyala/fasthttp"
+)
+
+func source() interface{} {
+	return make([]byte, 1)
+}
+
+func sink(interface{}) {
+}
+
+func fasthttpClient() {
+	userInput := "127.0.0.1:8909"
+	fasthttp.DialDualStack(userInput)           // $ SsrfSink=userInput
+	fasthttp.Dial(userInput)                    // $ SsrfSink=userInput
+	fasthttp.DialTimeout(userInput, 5)          // $ SsrfSink=userInput
+	fasthttp.DialDualStackTimeout(userInput, 5) // $ SsrfSink=userInput
+
+	req := &fasthttp.Request{}
+
+	uri1 := fasthttp.AcquireURI()
+	userInput = "UserControlled.com:80"
+	uri1.SetHost(source().(string))
+	sink(uri1) // $ hasTaintFlow="uri1"
+	uri2 := fasthttp.AcquireURI()
+	uri2.SetHostBytes(source().([]byte))
+	sink(uri2) // $ hasTaintFlow="uri2"
+	userInput = "http://UserControlled.com"
+	uri3 := fasthttp.AcquireURI()
+	uri3.Update(source().(string))
+	sink(uri3) // $ hasTaintFlow="uri3"
+	uri4 := fasthttp.AcquireURI()
+	uri4.UpdateBytes(source().([]byte))
+	sink(uri4) // $ hasTaintFlow="uri4"
+	uri5 := fasthttp.AcquireURI()
+	uri5.Parse(source().([]byte), source().([]byte))
+	sink(uri5) // $ hasTaintFlow="uri5"
```

Comment:
You need to test flow through each argument separately.
```suggestion
	uri5 := fasthttp.AcquireURI()
	uri5.Parse(source().([]byte), nil)
	sink(uri5) // $ hasTaintFlow="uri5"
	uri6 := fasthttp.AcquireURI()
	uri6.Parse(nil, source().([]byte))
	sink(uri6) // $ hasTaintFlow="uri6"
```